subxt/src/storage/storage_type.rs

```diff
@@ -290,6 +310,75 @@ where
     }
 }
 
+/// Return a vec of hashers and the associated type IDs for the keys that are hashed.
+fn storage_hasher_type_id_pairs(
+    entry: &StorageEntryType,
+    metadata: &Metadata,
+) -> Result<Vec<(StorageHasher, u32)>, Error> {
+    match entry {
+        StorageEntryType::Plain(_) => Ok(vec![]),
+        StorageEntryType::Map {
+            hashers, key_ty, ..
+        } => {
+            let ty = metadata
+                .types()
+                .resolve(*key_ty)
+                .ok_or(MetadataError::TypeNotFound(*key_ty))?;
+            match &ty.type_def {
+                TypeDef::Tuple(tuple) => {
+                    if hashers.len() < tuple.fields.len() {
+                        return Err(StorageAddressError::WrongNumberOfHashers {
+                            hashers: hashers.len(),
+                            fields: tuple.fields.len(),
+                        }
+                        .into());
+                    }
+                    let pairs: Vec<(StorageHasher, u32)> = tuple
+                        .fields
+                        .iter()
+                        .zip(hashers.iter())
+                        .map(|(e, h)| (*h, e.id))
+                        .collect();
+
+                    Ok(pairs)
+                }
+                _other => {
+                    if hashers.is_empty() {
+                        return Err(StorageAddressError::WrongNumberOfHashers {
+                            hashers: 0,
+                            fields: 1,
+                        }
+                        .into());
+                    }
+                    Ok(vec![(hashers[0], *key_ty)])
+                }
+            }
+        }
+    }
+}
+
+/// Strips the first 16 bytes (8 for the pallet hash, 8 for the entry hash) off some storage address bytes.
+fn strip_storage_addess_root_bytes(address_bytes: &mut &[u8]) -> Result<(), StorageAddressError> {
+    if address_bytes.len() >= 16 {
+        *address_bytes = &address_bytes[16..];
```

Comment:
this may end up with an empty slice if address_bytes.len == 16 ok?

Reply:
Yes, that is true, but it is okay I think.

Reply:
Yeah I think that's ok; if it matters (ie we expect a non-empty hash next) then the next thing will error instead